go/consensus/tendermint/registry/registry.go

```diff
@@ -189,72 +211,135 @@ func (tb *tendermintBackend) onEventDataNewBlock(ctx context.Context, ev tmtypes
 	events := append([]abcitypes.Event{}, ev.ResultBeginBlock.GetEvents()...)
 	events = append(events, ev.ResultEndBlock.GetEvents()...)
 
-	tb.onABCIEvents(ctx, events, ev.Block.Header.Height)
+	_, _ = tb.onABCIEvents(ctx, events, ev.Block.Header.Height, true)
 }
 
 func (tb *tendermintBackend) onEventDataTx(ctx context.Context, tx tmtypes.EventDataTx) {
-	tb.onABCIEvents(ctx, tx.Result.Events, tx.Height)
+	_, _ = tb.onABCIEvents(ctx, tx.Result.Events, tx.Height, true)
 }
 
-func (tb *tendermintBackend) onABCIEvents(ctx context.Context, events []abcitypes.Event, height int64) {
-	for _, tmEv := range events {
+func (tb *tendermintBackend) onABCIEvents(ctx context.Context, tmEvents []abcitypes.Event, height int64, doBroadcast bool) ([]api.Event, error) { // nolint: gocyclo
+	var events []api.Event
+	for _, tmEv := range tmEvents {
+		// Ignore events that don't relate to the registry app.
 		if tmEv.GetType() != app.EventType {
 			continue
 		}
 
 		for _, pair := range tmEv.GetAttributes() {
-			if bytes.Equal(pair.GetKey(), app.KeyNodesExpired) {
+			key := pair.GetKey()
+			val := pair.GetValue()
+			if bytes.Equal(key, app.KeyNodesExpired) {
+				// Nodes expired event.
 				var nodes []*node.Node
-				if err := cbor.Unmarshal(pair.GetValue(), &nodes); err != nil {
+				if err := cbor.Unmarshal(val, &nodes); err != nil {
 					tb.logger.Error("worker: failed to get nodes from tag",
 						"err", err,
 					)
+					if doBroadcast {
+						continue
+					} else {
+						return nil, fmt.Errorf("registry: corrupt NodesExpired event: %w", err)
+					}
 				}
 
-				for _, node := range nodes {
-					tb.nodeNotifier.Broadcast(&api.NodeEvent{
-						Node:           node,
-						IsRegistration: false,
-					})
+				if doBroadcast {
+					for _, node := range nodes {
+						tb.nodeNotifier.Broadcast(&api.NodeEvent{
+							Node:           node,
+							IsRegistration: false,
+						})
+					}
+				} else {
+					evt := api.Event{
+						NodesExpiredEvent: &api.NodesExpiredEvent{
+							Nodes: nodes,
+						},
+					}
+					events = append(events, evt)
+
+					// For compatibility, we also emit NodeEvents with
```

Comment:
Why have both?

Reply:
Because the rest of the code seems to expect these events and not the expiration events, especially in tests.  I can remove this if it doesn't make sense.

Reply:
Yeah but why have the separate expiration events at all?

Reply:
Oh, you meant that.  Don't know, I just exported all the events that the underlying layer had, maybe it will be useful to the developers that have requested this feature?

Reply:
But the two are basically equivalent (there is no other way to deregister a node, so node deregistration event is the expiration event) if you ignore aggregation (list vs. multiple events). I would rather have just one.

Reply:
OK, I'll remove the expiration event then :)

Reply:
Done!